substrate/frame/support/src/traits/metadata.rs

```diff
@@ -261,41 +261,53 @@ impl Add<u16> for StorageVersion {
-/// Otherwise it means that the pallet works with an unsupported storage version and unforeseen
-/// stuff can happen.
+/// Every pallet has two storage versions:
+/// 1. An current storage version
+/// 2. An on-chain storage version
 ///
-/// The current storage version is the version of the pallet as supported at runtime. The active
-/// storage version is the version of the pallet in the storage.
+/// The current storage version is the version of the pallet as defined in the local development
+/// environment, and the on-chain storage version is the version of the pallet actually deployed
+/// on-chain.
 ///
-/// It is required to update the on-chain storage version manually when a migration was applied.
+/// Storage versions should be only ever be out of sync when a pallet has been updated to a new
+/// version and the current version is incremented, but the migration has not yet been executed
+/// on-chain as part of a runtime upgrade.
+///
+/// It is the responsibility of the developer to ensure that the on-chain storage version is set
+/// correctly during a migration so that it matches the current storage version.
 pub trait GetStorageVersion {
-	/// This will be filled out by the [`pallet`](crate::pallet) macro.
+	/// This type is generated by the [`pallet`](crate::pallet) macro.
+	///
+	/// If the [`storage_version`](crate::pallet_macros::storage_version) attribute isn't specified,
+	/// this is set to [`NoStorageVersionSet`] to signify that it is missing.
+	///
+	/// If the [`storage_version`](crate::pallet_macros::storage_version) attribute is specified,
+	/// this is be set to a [`StorageVersion`] corresponding to the attribute.
 	///
-	/// If the [`storage_version`](crate::pallet_macros::storage_version) attribute isn't given
-	/// this is set to [`NoStorageVersionSet`] to inform the user that the attribute is missing.
-	/// This should prevent that the user forgets to set a storage version when required. However,
-	/// this will only work when the user actually tries to call [`Self::current_storage_version`]
-	/// to compare it against the [`Self::on_chain_storage_version`]. If the attribute is given,
-	/// this will be set to [`StorageVersion`].
+	/// The intention of using [`NoStorageVersionSet`] instead of defaulting to a [`StorageVersion`]
+	/// of zero is to prevent developers from forgetting to set
+	/// [`storage_version`](crate::pallet_macros::storage_version) when it is required, like in the
+	/// case that they wish to compare the current storage version to the on-chain storage version.
 	type CurrentStorageVersion;
 
-	/// Returns the current storage version as supported by the pallet.
+	/// Returns the current storage version as specified in the
+	/// [`storage_version`](crate::pallet_macros::storage_version) attribute, or
+	/// [`NoStorageVersionSet`] if the attribute is missing.
 	fn current_storage_version() -> Self::CurrentStorageVersion;
```

Comment:
I think the word "current" has always been slightly confusing to me here, maybe `in-code` would be better?

Reply:
I had been feeling 'current' was poor terminology for a while too but wasn't sure what was better, I really like 'in-code' :)  

Fairly big change, but updated code to use 'in-code' instead of 'current'.